torch/csrc/jit/script/schema_matching.cpp

```diff
@@ -417,22 +417,84 @@ MatchedSchema matchSchema(
     const SourceRange& loc,
     Graph& graph,
     at::ArrayRef<NamedValue> args,
-    at::ArrayRef<NamedValue> kwargs) {
+    at::ArrayRef<NamedValue> kwargs,
+    const c10::optional<NamedValue>& self) {
   std::stringstream failure_messages;
   if (auto result = tryMatchSchema(
           schema,
           loc,
           graph,
-          c10::nullopt,
           args,
           kwargs,
+          self,
           &failure_messages,
           /*allow_conversions=*/true)) {
     return *result;
   }
   throw ErrorReport(loc) << failure_messages.str();
 }
 
+static std::string prefixLine(
+    const std::string& str,
+    const std::string& prefix) {
+  std::stringstream ss;
+  bool was_newline = true;
+  for (auto c : str) {
+    if (was_newline)
+      ss << prefix;
+    ss.put(c);
+    was_newline = c == '\n';
+  }
+  return ss.str();
+}
+
+std::pair<size_t, MatchedSchema> matchSchemas(
+    const std::vector<const FunctionSchema*>& schemas,
+    const SourceRange& loc,
+    Graph& graph,
+    at::ArrayRef<NamedValue> args,
+    at::ArrayRef<NamedValue> kwargs,
+    const c10::optional<NamedValue>& self,
+    bool render_errors) {
+  TORCH_INTERNAL_ASSERT(schemas.size() > 0);
+  if (schemas.size() == 1) {
+    return std::make_pair(
+        0, matchSchema(*schemas.at(0), loc, graph, args, kwargs, self));
+  }
+  std::stringstream failure_messages;
+  for (bool allow_conversions : {false, true}) {
+    // clear previous error messages
+    failure_messages.str("");
+    for (size_t i = 0; i < schemas.size(); ++i) {
+      const auto matched_schema = tryMatchSchema(
+          *schemas[i],
+          loc,
+          graph,
+          args,
+          kwargs,
+          self,
+          render_errors ? &failure_messages : nullptr,
+          allow_conversions);
+      if (matched_schema) {
+        return std::make_pair(i, std::move(*matched_schema));
+      }
+    }
+  }
+  // we optimistically assume this call will not error, and avoid formatting the
+  // error strings. If we discover it did error, then we replay it, recording
+  // the errors.
+  if (!render_errors) {
+    return matchSchemas(
```

Comment:
i don't understand this - if we have reached this point in the code, none of the schemas match, can't we just throw the error ?

Reply:
We start with render_errors=false, optimistically assuming that we will succeed and not spending time doing error message formatting for an overload that won't happen. If we reach here we know there will be an error, but we have not rendered the error messages, so we replay the match to render the errors.

Reply:
I assume this is just to cut down on compile time. It would be nice to hide the `render_errors` argument from the api since any invocation outside of this file shouldn't need to know about this. 

Related: we ran into a compile - slow down in compiler.cpp with eagerly concatenating strings and it was solved by buffering up the concatenation with lambdas: https://github.com/pytorch/pytorch/blob/264c4380b868fd985fd91cc6a6a16995287c38af/torch/csrc/jit/script/compiler.cpp#L236